[#50] 팔로우 알람 push

choitaehoon opened the pull request:
- 파이어베이스 message 전송

choitaehoon committed:
[#11] 게시물 작성

- 게시물 작성 구현
- AccountController, LoginController naming 변경

choitaehoon committed:
[#14] AWS S3 환경설정

- AWS의 key값 분리 후 gitignore 적용
- AmazonS3 빈으로 등록

choitaehoon committed:
[#14] AWS S3 환경설정

- javax.annotation-api 종속성 추가

choitaehoon committed:
[#14] AWS S3 환경설정

- PropertySource 제거 후 EnvironmentPostProcessor 생성
- aws 환경설정 예외 처리
- aws 환경 설정, 메시지에 관련된 상수 생성

choitaehoon committed:
[#14] AWS S3 환경설정

- gitignore application-aws.yml로 변경

choitaehoon committed:
[#14] AWS S3 환경설정

- 주석제거

choitaehoon committed:
[#14] AWS S3 환경설정

- EnvironmentPostProcessor 제거
- application.yml 파일안에 aws include

choitaehoon committed:
[#16] Aws S3 이미지 업로드

- AWS S3 이미지 업로드 구현
- 파일 예외 처리

choitaehoon committed:
[#16] Aws S3 이미지 업로드

- FileController API naming 변경
- upload 상수로 반환
- 에러 로그 추가
- AWS S3 인터페이스로 확장

choitaehoon committed:
[#20] 게시물에 대한 전반적인 도메인 생성

- 게시물, 댓글, 좋아요, 좋아요 클릭 유저 Entity 생성
- BaseTimeEntity class 생성하여 생성 날짜, 수정 날짜 자동으로 변경

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- AwsS3Config 안에 bucket 빈으로 등록
- 예외 에러 로그 추가
- AwsS3Connector에서 AwsS3LocalService로 upload 옮기기

choitaehoon committed:
[#20] 게시물에 대한 전반적인 도메인 생성

- Post 클래스 fileName 제거

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- ExceptionHandler를 이용하여 예외 객체 로그 적용
- upload 로그 debug로 낮추기

choitaehoon committed:
[#20] 게시물에 대한 전반적인 도메인 생성

- 생성자 생성

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- Post Entity에 filePath 추가
- S3 이미지 업로드 후 Post Entity에 filePath, userId 저장

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- PostEntity로 변경하는 메소드 naming 변경

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- 디버깅에 용이하게 만들고자 ExceptionHandler안에 stacktrace 출력 구현

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- Exception Handler error 로그 수정
- AwsS3LocalService @Value 생성자 주입
- AwsS3Config bucket 빈 삭제

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- ExceptionUtils 의존성 삭제

choitaehoon committed:
Merge branch 'develop' into feature/16-aws-upload

choitaehoon committed:
[#23] 피드 조회

- 이미지 업로드 이후 comment 수정
- 피드 조회
- application-local.yml logging 한줄로 변경

choitaehoon committed:
Merge branch 'feature/16-aws-upload' into feature/20-post-and-nice-domain

choitaehoon committed:
Merge branch 'feature/20-post-and-nice-domain' into feature/23-feeds-select

choitaehoon committed:
[#23] 피드 조회와 피드 작성 이후 comment 수정

- PostService updateComment 매개변수 수정
- Mysql 예악어 대문자로 변경

choitaehoon committed:
[#23] 피드 조회와 피드 작성 이후 comment 수정

- PostResponseDto 생성 및 반환
- PostRepository 메소드명 변경

choitaehoon committed:
[#23] 피드 조회와 피드 작성 이후 comment 수정

- 피드 조회 페이징 처리

choitaehoon committed:
[#23] 피드 조회와 피드 작성 이후 comment 수정

- PostResponseDto 클래스 안에 있는 NoArgsConstructor 제거

choitaehoon committed:
[#26] 레디스 CacheManager 설정

- 레디스 CacheManager 설정 후 빈으로 등록

choitaehoon committed:
[#23] 피드 조회와 피드 작성 이후 comment 수정

- 특정 유저 피드 조회 시 query string으로 변경

choitaehoon committed:
[#26] 레디스 CacheManager 설정

- 세션과 캐쉬 레디스 분리
- RedisConfig -> RedisSessionConfig로 naming 변경

choitaehoon committed:
[#28] 특정 유저 피드 조회 캐싱 적용

- Cacheable 어노테이션으로 특정 유저 피드 조회 캐싱

choitaehoon committed:
[#28] 특정 유저 피드 조회 캐싱 적용

- redis cache 만료기간 설정
- PostService에 캐쉬 적용

choitaehoon committed:
[#28] 특정 유저 피드 조회 캐싱 적용

- redis cache 만료기간 naming 변경

choitaehoon committed:
[#28] 특정 유저 피드 조회 캐싱 적용

- redis cache 만료기간 naming 변경

choitaehoon committed:
[#28] 특정 유저 피드 조회 캐싱 적용

- RedisCacheConfiguration Map타입으로 여러개 시간 설정 할 수 있도록 구현

choitaehoon committed:
[#28] 특정 유저 피드 조회 캐싱 적용

- FEEDS_CACHE_EXPIRE_MINUTE 변수명앞에 USER_ID 붙이기

choitaehoon committed:
[#28] 특정 유저 피드 조회 캐싱 적용

- 도메인 관심사를 분리하기 위해 Feeds naming으로 변경

choitaehoon committed:
[#28] 특정 유저 피드 조회 캐싱 적용

- 캐시 만료, 이름 외부에 주입
- userIdFeed -> feedsPerUser로 naming 수정

choitaehoon committed:
[#32] 특정 유저 피드 조회할 때 좋아요 칼럼 추가

- PostResponseDto nice 필드 추가
- PostNiceRepository를 이용하여 nice 필드 조회

choitaehoon committed:
[#34] 피드 좋아요 클릭

- 피드 좋아요 클릭시 update 처리

choitaehoon committed:
[#32] 특정 유저 피드 조회할 때 좋아요 칼럼 추가

- PostResponseDto 메소드 이름 from으로 변경

choitaehoon committed:
[#36] 유저 피드 조회 캐싱할 때 좋아요 제거

- 유저 피드 조회 캐싱에서 좋아요 제거 완료

choitaehoon committed:
[#38] 카프카 설정

- Producer, Consumer, Topic 설정잡기

choitaehoon committed:
[#38] 카프카 설정

- Producer, Consumer, Topic 설정 변수 naming 변경

choitaehoon committed:
[#38] 카프카 설정

- batch, topic 설정

choitaehoon committed:
[#40] 카프카 배치 처리

- KafkaListener 구현 (아직 batch에 대한 이해도가 높지 않아 공부하면서 세밀하게 구현할 예정)

choitaehoon committed:
[#40] 카프카 배치 처리

- kafkaListener topic, groupId SpEL 식으로 변경

choitaehoon committed:
[#40] 카프카 배치 처리

- kafkaListener 어노테이션 설명

choitaehoon committed:
[#40] 카프카 배치 처리

- kafkaListener 어노테이션 설명 보완

choitaehoon committed:
[#42] 팔로워, 팔로우 도메인 생성

- Follower, Follow 클래스 생성

choitaehoon committed:
[#44] 팔로우  구현

- 팔로우 service 구현

choitaehoon committed:
[#42] 팔로우 도메인 생성

- Follower 도메인 삭제

choitaehoon committed:
[#44] 팔로우 구현

- 팔로우 API, service 구현

choitaehoon committed:
[#44] 팔로우 구현

- sendFollow -> follow로 변경

choitaehoon committed:
[#44] 팔로우 구현

- HTTP 상태코드 201로 변경

choitaehoon committed:
[#44] 팔로우 구현

- HTTP 상태코드 204로 변경

choitaehoon committed:
[#46] 팔로우 취소

- 팔로우 delete 구현

choitaehoon committed:
[#46] 팔로우 취소

- SQLException 예외를 FollowNotDeleteException으로 감싸서 전달하기

choitaehoon committed:
[#46] 팔로우 취소

- 팔로우 예외 naming 변경
- SQLException을 DataAccessException으로 변경

choitaehoon committed:
[#44] 팔로우 구현

- HttpStatus 204 -> 201로 변경

choitaehoon committed:
[#46] 팔로우 취소

- HttpStatus 204 -> 201로 변경

choitaehoon committed:
[#48] 팔로우 알람 push 설정

- 파이어베이스 연동
- 파이어베이스 연동 실패 예외 작성

choitaehoon committed:
[#50] 팔로우 알람 push

- 파이어베이스 message 전송

f-lab-dev commented on src/main/java/com/social/instagram/service/FollowService.java:
```diff
@@ -1,22 +1,49 @@
     public void follow(Follow follow) {
         followRepository.save(follow);
+
+        FirebaseMessaging.getInstance().sendAsync(followMessage(follow));
```
`푸시 메세지를 전송한다`라는 행위는 캡슐화가 필요해보입니다. 푸시메세지 전송에 어떤 비지니스 로직이 추가될지 모르는데 이렇게 바로 갖다쓰면 추후 유지보수성이 떨어지지 않을까요?

f-lab-dev commented on src/main/java/com/social/instagram/service/FollowService.java:
```diff
@@ -1,22 +1,49 @@
 
+    private Message followMessage(Follow follow) {
+        return Message.builder()
+                .setToken(firebaseToken)
+                .setNotification(Notification.builder()
+                        .setTitle(messageTitle)
+                        .setBody(follow.getUserId() + messageBody)
+                        .build())
+                .build();
+    }
```
이 메소드는 팔로우에 관련된 작업일까요? 푸시메세지에 관련된 작업일까요?

choitaehoon replied:
푸쉬메시지에 관한 작업을 `FollowService`에서 구현했었네요. 분리하겠습니다!

choitaehoon committed:
[#50] 팔로우 알람 push

- 팔로우와 푸시메시지 관심사 분리

f-lab-dev commented on src/main/java/com/social/instagram/service/FirebaseService.java:
```diff
@@ -0,0 +1,39 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class FirebaseService {
```
이 서비스는 `푸시 메시지를 보낸다`로 추상화가 가능해보이네요

f-lab-dev commented on src/main/java/com/social/instagram/service/FirebaseService.java:
```diff
@@ -0,0 +1,39 @@
+        this.messageBody = messageBody;
+    }
+
+    public void sendAsyncMessage(Follow follow) {
```
비동기로 보내는걸 명시적으로 나타낸다면 이 메소드를 사용하는 개발자들이 별도의 콜백을 추가할 수 있게 `Future`를 리턴해주는게 좋습니다

choitaehoon replied:
좋은 정보 감사합니다! `Future`에 대해서도 공부해보겠습니다

choitaehoon committed:
[#50] 팔로우 알람 push

- FirebaseService -> PushMessageService로 naming 변경
- 추후에 비동기 작업에 callback을 추가할 수 있게 Future 리턴

f-lab-dev commented on src/main/java/com/social/instagram/service/PushMessageService.java:
```diff
@@ -7,23 +7,25 @@
+
 @Service
-public class FirebaseService {
+public class PushMessageService {
```
추상화를 한다는 것은 구체적인 타입이 아니라 인터페이스로 뺀다는 것입니다~ 그래야 상속받아서 IOS 푸쉬도 날릴 수 있겠죠

choitaehoon replied:
아 제가 착각했네요. 인터페이스로 빼서 `OCP원칙`과 `전략 패턴`을 적용할 수 있게 하겠습니다!

choitaehoon committed:
[#50] 팔로우 알람 push

- 파이어베이스 인터페이스로 분리

choitaehoon commented on src/main/java/com/social/instagram/service/firebase/WebPushMessageService.java:
```diff
@@ -0,0 +1,42 @@
+        return FirebaseMessaging.getInstance().sendAsync(from(follow));
+    }
+
+    private Message from(Follow follow) {
```
현재 메시지를 만드는 기능을 `WebPushMessageService`에만 접근 할 수 있게 구현되어 있습니다. `IOS`나 `Android`는 이 기능을 사용하지 못하는 문제점을 파악해서 이 부분을 해결하겠습니다

f-lab-dev commented on src/main/java/com/social/instagram/service/firebase/FirebaseService.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.concurrent.Future;
+
+public interface FirebaseService {
```
이름도 추상화시키는게 좋을 것 같습니다~ `Firebase`는 안드로이드에 종속적인게 아닐까요?

choitaehoon replied:
아 그렇군요. 이름 변경하겠습니다

choitaehoon committed:
[#50] 팔로우 알람 push

- 파이어베이스 인터페이스, 서비스 naming 변경
- MessageService 추출

f-lab-dev commented on src/main/java/com/social/instagram/service/firebase/WebPushService.java:
```diff
@@ -0,0 +1,22 @@
+
+@Service
+@RequiredArgsConstructor
+public class WebPushService implements PushService {
```
이 클래스는 파이어베이스로 푸시를 보내는 클래스 같네요. 그러면 `MessageService`와 다른 점을 모르겠습니다.
그 곳에서 쓰고있는 `Message` 클래스 또한 파이어베이스 종속적입니다.

`MessageService`는 파이어베이스 메세지를 만들어주는 역할로 분리하신거라면 `PushMessageBuilder` 혹은 `PushMessageFactory`가 낫지 않을까요?

choitaehoon replied:
`MessageService`를 파이어베이스 메시지 만드는 역할로 분리했습니다. 단순히 메시지를 만들고 반환하는 용도로 사용하니 `PushMessageFactory`가 적합할 것 같네요

choitaehoon committed:
[#50] 팔로우 알람 push

- 메시지 생성을 Factory로 분리

f-lab-dev commented on src/main/java/com/social/instagram/service/firebase/WebPushService.java:
```diff
@@ -12,11 +12,11 @@
     @Override
     public Future<String> sendAsyncMessage(Follow follow) {
-        return FirebaseMessaging.getInstance().sendAsync(messageService.from(follow));
+        return FirebaseMessaging.getInstance().sendAsync(pushMessageFactory.from(follow));
```
`from`이란 메소드는 인자로부터 자신의 객체을 만든다는 의미로 자주 쓰입니다. `getMessage()` 같이 좀 더 명확하게 이름을 지어주면 좋을 것 같네요~

choitaehoon replied:
`from`의 이름은 도메인에서 사용할 때 어울리겠네요. `getMessage()`로 변경하겠습니다

choitaehoon committed:
[#50] 팔로우 알람 push

- from -> getMessage naming 변경

choitaehoon committed:
Merge branch 'develop' into feature/50-follow-push
